Fix World#isChunkGenerated calls

Spottedleaf opened the pull request:
Optimize World#loadChunk() too
This patch also adds a chunk status cache on region files (note that
its only purpose is to cache the status on DISK)

Note that I still need to review this with a plugin, however debug reports show the concept works.
This fixes upstream issue https://hub.spigotmc.org/jira/browse/SPIGOT-4862

Spottedleaf committed:
Fix World#isChunkGenerated calls

Optimize World#loadChunk() too
This patch also adds a chunk status cache on region files (note that
its only purpose is to cache the status on DISK)

blablubbabc commented on Spigot-Server-Patches/0403-Fix-World-isChunkGenerated-calls.patch:
```diff
@@ -0,0 +1,312 @@
+-            world.getChunkProvider().addTicket(TicketType.PLUGIN, new ChunkCoordIntPair(x, z), 1, Unit.INSTANCE);
+-            return true;
++        // Paper - Optimize this method
++        if (generate && !isChunkGenerated(x, z)) {
```
Was probably meant to be
```suggestion
+        if (!generate && !isChunkGenerated(x, z)) {
```
Same below

Spottedleaf replied:
I'll get the other one soon

Spottedleaf committed:
Update Spigot-Server-Patches/0403-Fix-World-isChunkGenerated-calls.patch

yeah oops

Co-Authored-By: blablubbabc <blablubbabc@users.noreply.github.com>

blablubbabc commented on Spigot-Server-Patches/0403-Fix-World-isChunkGenerated-calls.patch:
```diff
@@ -0,0 +1,312 @@
+-        if (chunk instanceof net.minecraft.server.Chunk) {
+-            return ((net.minecraft.server.Chunk)chunk).bukkitChunk;
++        // Note: Copied from loadChunk()
++        if (gen && !isChunkGenerated(x, z)) {
```
```suggestion
+        if (!gen && !isChunkGenerated(x, z)) {
```

Spottedleaf committed:
Update Spigot-Server-Patches/0403-Fix-World-isChunkGenerated-calls.patch

oops part 2

Co-Authored-By: blablubbabc <blablubbabc@users.noreply.github.com>

Spottedleaf committed:
Ensure correct regionfile usage

This also bumps the minimum region file cache to 4 files given
readChunkData can load potentially 4 files.

Spottedleaf committed:
Fix closed check

Spottedleaf committed:
Better checks for invalid regionfile usage

Spottedleaf committed:
fix position of comment
